pkg/serializer/func_serializer.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serializer
+
+var (
+	noRetry = func(i int) bool { return false }
+)
+
+type funcProp struct {
+	f        func() error
+	waitFunc func(int) bool
+}
+
+type funcSerializer struct {
+	funcs chan funcProp
+}
+
+// NewFuncSerializer returns a funcSerializer that will be used to execute
+// functions in the same order they are enqueued.
+func NewFuncSerializer(size uint) *funcSerializer {
+	return &funcSerializer{
+		funcs: make(chan funcProp, size),
+	}
+}
+
+// Run starts the funcSerializer internal worker. It will be stopped once
+// `stopCh` is closed or receives a value.
+func (fs *funcSerializer) Run(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		case f := <-fs.funcs:
+			retries := 0
+			for {
+				select {
+				case <-stopCh:
+					return
+				default:
+				}
+				retries++
+				if err := f.f(); err != nil {
+					if !f.waitFunc(retries) {
+						break
+					}
+				} else {
+					break
+				}
+			}
+		}
+	}
+}
+
+// Enqueue enqueues the receiving function `f` to be executed by the function
+// serializer. Depending on the size of the function serializer and the amount
+// of functions queued, this function can block until the function serializer
+// is ready to receive more requests.
+// The waitFunc will be executed with the number of times `f` had returned an
+// error, in case `f` returns an error. waitFunc should return either true or
+// false depending if `f` should be executed again or not.
+func (fs *funcSerializer) Enqueue(f func() error, waitFunc func(int) bool) {
```

Comment:
Please define a documented type for `waitFunc` which explains what `int` will be and what it has to return

Reply:
Done.